src/Compilers/CSharp/Portable/Compilation/CSharpDiagnosticFilter.cs

```diff
@@ -70,11 +71,18 @@ public static Diagnostic Filter(Diagnostic d, int warningLevelOption, ReportDiag
                     d.Category,
                     warningLevelOption,
                     generalDiagnosticOption,
-                    specificDiagnosticOptions);
+                    specificDiagnosticOptions,
+                    out hasPragmaSuppression);
             }
             else
             {
-                reportAction = GetDiagnosticReport(d.Severity, d.IsEnabledByDefault, d.Id, d.WarningLevel, d.Location as Location, d.Category, warningLevelOption, generalDiagnosticOption, specificDiagnosticOptions);
+                reportAction = GetDiagnosticReport(d.Severity, d.IsEnabledByDefault, d.Id, d.WarningLevel, d.Location as Location,
+                    d.Category, warningLevelOption, generalDiagnosticOption, specificDiagnosticOptions, out hasPragmaSuppression);
+            }
+
+            if (hasPragmaSuppression)
+            {
+                d = d.WithHasSourceSuppression(true);
```

Comment:
Don't we also want to set "has source suppression" if there was a `SuppressMessage` attribute? (And what do we do if the `SuppressMessage` attribute is in the sidecar file (GlobalSuppressions.cs)?